tests/host_tools/network.py

```diff
@@ -52,7 +54,17 @@ def __init__(self, netns, ssh_key: Path, host, user):
             str(self.ssh_key),
         ]
 
-        self._init_connection()
+        # _init_connection loops until it can connect to the guest
+        # dumping debug state on every iteration is not useful or wanted, so
+        # only dump it once if _all_ iterations fail.
+        try:
+            self._init_connection()
+        except Exception as exc:
+            on_error(exc)
```

Comment:
what if `on_error` is `None`?

We could define `on_error` as `lambda x: print(x)`

Reply:
argh, I handled the none case below, but forgot here. Thanks!

If we define it as `lambda x: print(x)` we'll end up just printing the exception twice, without attaching any new information to it, so I'd prefer to not do that (although `on_error` is always set in current usecases anyway, so probably it does not matter).